lib/debug/thread_client.rb

```diff
@@ -344,29 +344,43 @@ def instance_eval_for_cmethod frame_self, src
       frame_self.instance_eval(src)
     end
 
+    def eval_binding
+      current_frame&.binding || TOPLEVEL_BINDING
+    end
+
     SPECIAL_LOCAL_VARS = [
       [:raised_exception, "_raised"],
       [:return_value,     "_return"],
     ]
+    SPECIAL_LOCAL_VAR_NAMES = SPECIAL_LOCAL_VARS.map { |_, n| n }
 
     def frame_eval src, re_raise: false
       @success_last_eval = false
 
-      b = current_frame&.eval_binding || TOPLEVEL_BINDING
-      b = b.dup
+      # the binding of original program
+      original_binding = eval_binding
+      # the binding used by debugger REPL that can have special locals
+      repl_binding = eval_binding.dup
 
-      special_local_variables current_frame do |name, var|
-        b.local_variable_set(name, var) if /\%/ !~ name
+      special_local_variables current_frame, binding: repl_binding do |name, var|
+        repl_binding.local_variable_set(name, var) if /\%/ !~ name
       end
 
-      result = if b
-                  f, _l = b.source_location
-                  b.eval(src, "(rdbg)/#{f}")
+      result = if repl_binding
+                  f, _l = repl_binding.source_location
+                  repl_binding.eval(src, "(rdbg)/#{f}")
                 else
                   frame_self = current_frame.self
                   instance_eval_for_cmethod(frame_self, src)
                 end
       @success_last_eval = true
+
+      # write non-special local variables back to the original binding
+      repl_binding.local_variables.each do |var|
+        next if SPECIAL_LOCAL_VAR_NAMES.include?(var)
+        original_binding.local_variable_set(var, repl_binding.local_variable_get(var))
+      end
```

Comment:
After we dup the original binding, local variables changes made in the REPL won't be applied to it. So we need to manually propagate the values back to the original binding.

Reply:
on master, it is propagated. which case do you considering?

```
[1, 5] in target.rb
=>   1| a = b = c = nil
     2| p a, b, c
     3|
     4|
     5| __END__
=>#0    <main> at target.rb:1
(rdbg) n    # next command
[1, 5] in target.rb
     1| a = b = c = nil
=>   2| p a, b, c
     3|
     4|
     5| __END__
=>#0    <main> at target.rb:2
(rdbg) eval a = b = c = :ok    # command
(rdbg) n    # next command
:ok
:ok
:ok
```

Reply:
maybe you mean:

```
[1, 4] in target.rb
     1|
=>   2| a = 1
     3|
     4| __END__
=>#0    <main> at target.rb:2
(rdbg) y = 50    # ruby
50
(rdbg) y
eval error: undefined local variable or method `y' for main:Object
  (rdbg)/target.rb:1:in `<main>'
nil
```

but `y` is only valid on this debug subsession (when continue, the `y` will be gone). hmm

Reply:
byebug's example.

```ruby
[master]$ byebug target.rb

[1, 10] in /mnt/c/ko1/src/rb/ruby-debug/target.rb
    1:
=>  2: a = 1
    3: b = 2
    4:
    5: __END__
    6:
    7: require 'foo'
    8:
    9: __END__
   10:
(byebug) y = 10
10
(byebug) p y
10
10
(byebug) n

[1, 10] in /mnt/c/ko1/src/rb/ruby-debug/target.rb
    1:
    2: a = 1
=>  3: b = 2
    4:
    5: __END__
    6:
    7: require 'foo'
    8:
    9: __END__
   10:
(byebug) y
*** NoMethodError Exception: undefined method `end_line=' for nil:NilClass
```